Add new tutorial for multi-tenant model [WIP]

samay-sharma opened the pull request:
Things left to do:

- [x] Upload data and link to it.
- [x] Figure out what we have to do for copying data with docker setup.
- [x] Decide on whether we want to add a shutdown section.
- [x] Add note about copy from different path.
- [x] Decide if we want to show query results or make it fully copy-pastable.

samay-sharma commented:
@ozgune : Still have a few TODOs, major ones are already listed above but I feel it is in a state to be looked at.

Could you have a look at it and provide feedback?

ozgune commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+
+::
+                                                                                          
+    \copy companies from 'companies.csv' WITH CSV;                                                     
```
@samay-sharma - I spinned up a Docker image and started going through these steps.

Where can I download these sample CSV files?

samay-sharma replied:
I added a step for `docker cp`. For own `psql`, I added a note for using the correct path. I could also change it to `PATH_TO_FILE/companies.csv` if you think that would make it better. I'll let you review and decide on that.

ozgune commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+                                                                                          
+::
+
+    SELECT count(*) from companies;                                                           
```
What do you think about swapping one of the following three ``SELECT`` statements with ``INSERT``s?

That way, our tutorial shifts some emphasis from the OLAP to the OLTP side?

samay-sharma replied:
Added an `INSERT` command and moved the OLTP part to the top.

ozgune commented on installation/production_deb.rst:
```diff
@@ -104,7 +104,7 @@ shell should output the worker nodes we added to the pg_dist_node table above.
 **Ready to use Citus**
 
-At this step, you have completed the installation process and are ready to use your Citus cluster. To help you get started, we have a :ref:`tutorial<tut_hash>` which has instructions on setting up a Citus cluster with sample data in minutes.
+At this step, you have completed the installation process and are ready to use your Citus cluster. To help you get started, we have a :ref:`tutorial<multi_tenant_tutorial>` which has instructions on setting up a Citus cluster with sample data in minutes.
```
@samay-sharma / @begriffs -- I had two questions.

1. If we link to the tutorial from Debian / RHEL pages, what do you think about adding steps for "localhost" in a future iteration?
2. Why do we point to our tutorial from production install pages but not from others (including Docker)?

ozgune replied:
I captured the points in this issue in #293.

samay-sharma replied:
@begriffs addressed this with his PR.

ozgune commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+                                                                                             
+    This tutorial assumes that you already have Citus installed and running. If you don't have Citus running,
+    you can provision a cluster using `Citus Cloud <https://console.citusdata.com>`_ or setup Citus locally
+    using :ref:`single_machine_docker`.
```
Unrelated question on the Docker install page.

@jasonmp85, @begriffs -- I had quick questions on a few discrepancies between our Docker setup installations on our docs and Github readme. @sumedhpathak -- I'm planning to open an issue on this, does including the following in that issue work?

* We point to Docker Compose in our Readme. We give the user the option to install Docker Engine or Docker Compose in our docs. Which one should we recommend?
* We ask the user to pick docker-compose for Linux or Mac and also ask them to run `eval`. We don't do that in our docs. Why is that?
* We ask the user to shut down Docker at the end. If they do and then come to this page, do we break the flow?

ozgune replied:
@samay-sharma -- I opened #292 to cover two of the three bullet points in this comment.

We still ask the user to shut down Docker in our docs. If we point them to the Docker page, it may introduce a minor flow issue. Are we worried about that at all -- enough to capture it some place?

samay-sharma replied:
@begriffs addressed this in his PR. Regarding the shut down and flow issue, let me see how best to address it.

samay-sharma replied:
I added a link to tutorial before shutdown in the docker instructions. I'm not planning to add shutdown instructions to the tutorial itself.

@ozgune Let me know if you'd like to add that.

ozgune commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+
+::                                                                                          
+    
+    UPDATE campaigns
```
This command only modifies a single row. What if we updated for all records for company_id = 5?

samay-sharma replied:
Changed to:

```sql
 UPDATE campaigns                                                                      
 SET monthly_budget = monthly_budget*2                                                 
 WHERE company_id = 5;
```

samay-sharma commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+::
+
+    CREATE TABLE companies (                                                                     
+        id integer NOT NULL,                                                                     
```
Maybe use `bigint` instead of `integer`?

samay-sharma replied:
Done

ozgune commented:
@samay-sharma -- I had a higher level question. We previously thought of tutorials as the simplest way to get started and therefore put them ahead of Single Node Installation steps.

What do we envision this tutorial is going to end up?

1. We could point you to the Docker installation page and then add a link so that you could come back to the tutorial. In that case, we should probably move this section to after Single Node Install.
2. We let you pick the platform you want to test things on. We then render quick installation steps for you on the fly. In that case, this section could probably stay where it's at.

begriffs commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+
+Then, you can create the tables by using standard PostgreSQL :code:`CREATE TABLE` commands.
+
+::
```
Small note - the code blocks render better in sphinx if you label them with a language. In this case you could do

```rst
.. code-block:: sql
```

rather than the bare double colon. Syntax highlighting is supported for [these languages](http://pygments.org/languages/).

samay-sharma replied:
Discussed in person with @begriffs . We already have a `.. highlight:: sql` at the top. So, this isn't needed.

aamederen commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,166 @@
+and allow for features like primary keys, foreign keys and complex joins across your cluster.
+You can learn more about the benefits of this approach `here <https://www.citusdata.com/blog/2016/10/03/designing-your-saas-database-for-high-scalability/>`_.
+                                                                                          
+We could now go ahead and load data into these tables. You can download sample files for all the tables here:
```
Are the files uploaded to somewhere (s3 maybe?)

samay-sharma replied:
I have uploaded them to s3. Will update the PR today with links.

aamederen commented:
I want to drop a note here. 

Depending on data size, it may take several minutes for the users to download and copy the data. This can lead to losing attention of the readers. 

One of the solution ideas for this problem can be providing citus clusters on the cloud with the data inside them (as we have done in PGConfSV 2016 tutorial), either in local tables or data files in the node servers. This may reduce the time to complete the tutorial.

samay-sharma commented:
@ozgune : Regarding the order of Single-node Install vs Tutorial, what you said makes sense. Given the current setup, I've moved the tutorial to after single node install. Once we add the dynamic instructions based on platform, we can revisit where want to put the tutorial.

samay-sharma commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,184 @@
+.. note::
+
+    Docker users can then copy the files into the docker container in order to load them.
+    You can do so by running the docker cp command and then connecting back into the server.
```
(Minor) Highlight docker cp as it is a command.

samay-sharma commented on tutorials/multi-tenant-tutorial.rst:
```diff
@@ -0,0 +1,184 @@
+
+::
+                                                                                          
+    \copy companies from 'companies.csv' WITH CSV;                                                     
```
Capitalize `\copy` and `from`

jasonmp85 replied:
:-1: for this. I'd actually say lowercase `with` and `csv`. One of the things that drove me crazy about Citus in the early days was how it consistently capitalized `\STAGE`. PostgreSQL's documentation consistently lowercases client-side `psql` commands and uppercases SQL keywords. By writing `copy`, `from`, and `with csv` as lowercase, it makes it more obvious to the reader they're `psql` commands.

jasonmp85 replied:
(See [the `psql` documentation](https://www.postgresql.org/docs/9.6/static/app-psql.html#APP-PSQL-META-COMMANDS-COPY) for more)

samay-sharma commented:
@ozgune : Moved the downloading data and `docker cp` command to the top. Also, changed title to Multi Tenant Applications. Let me know if it looks good to go.

ozgune commented:
@samay-sharma -- this looks good to me. Let's 🚀

begriffs commented:
Should I cherry-pick this into v6.1 as well?

sumedhpathak commented:
I don't see a reason not to. @samay-sharma @ozgune ?
